### DISCUSSION TITLE
remove or hide __next_data__?

### DISCUSSION THREAD

**Discussion by cakerake** [vercel/next.js#13418] [Help] [answered]
I'm using getinitial props to fetch data and then rendering a page, every page keeps getting a script tag with the entire json object just pasted into the html.

Is there a way to keep this json hidden and completely render the page on the server instead?

**Comment by jamesmosier** [comment, marked as answer]
That is how Next knows what data to render on the server, otherwise it wouldn't know what the data is that React should use to hydrate your components. It is not removable. 

The data contained in there should be what is already presented to your user, so no hidden or sensitive information should be leaked there.

**Comment by viktor32** [comment]
Hello
@jamesmosier thanks for the answer. But what about the situation when we use `memo` / `Pure` components and `getStaticProps`?

We have the next situation:

1. content data loaded as rendered HTML
2. the same content data included into the same HTML page
3. and the same content data downloaded as a separate JSON file

and that’s all when you first load the page.

But if we have page as memo component we know that the page will not rerendered if data do not change.  And if we use `getStaticProps` we know that data will not be changed. So can we say that during first loading the page we can load only HTML without data in `__next_data__ ` and json file? Of course, when we go to another page we have to download json data to render the next page. I wonder if this is possible?

**Comment by JakubKoralewski** [comment]
I have a statically generated blog. I put the data into HTML and now I found out that the content is is put **twice** into the HTML. Waste of data.

**Reply by AmrSaber** [reply]
This also increases the content size of the served page.

**Comment by AmrSaber** [comment]
@jamesmosier I use localization in my app, in the localization, I use a dictionary object that contains the localized string for each part of my application, this dictionary is passed to the page component using `getInitialProps`, and it's no longer needed when the page is rendered.

Now there are 2 problems:
1. -Almost- All the content of the page is served twice (once inside the page, and another time because the dictionary object is sent to the client)
1. Some additional strings that were not used in the page are also sent along with the dictionary that I used.

Isn't there a way to remove that dictionary from the served page after it has been rendered on the server?

**Reply by jamesmosier** [reply]
In order for Next (well really React) to determine what data to inject into the DOM when server rendering it needs to get that data from somewhere, which in this case is the `__next_data__` DOM element. 

If there is data not needed by the page, don't return it from `getInitialProps` (or any of the other data fetching methods). 

Any performance issue with your site is likely not caused by this. This is such a small piece of application performance that there are a ton of other places that are better to look at than some extra content in the DOM.

**Reply by JakubKoralewski** [reply]
tbh I don't really know how React or Next works exactly in this case, but whatever they do it is clearly wrong. I understand when the props are a piece of json array data that is then transformed into some elements, but what I believe many of us here are talking about is simply replacing some identifier (object property access etc.) with a piece of text. SSG already replaces the identifiers rendering the json data useless. I feel like a Babel plugin replacing some strings could be more efficient than the current state.

Furthermore while I appreciate the volunteer work of the open source community this project included, it is in my opinion incorrect and a bit shallow to dismiss this issue as there is a difference in having little duplicate data in a single page, and having a framework used by thousands that is contributing to a lot of useless data transmission and slower load times. Also Id like to point out how blogs can be built with this framework where there could be very long posts and a lot of posts on a single webpage. You're right this is a small problem in comparison, and there are probably more valuable things to work on, but as a framework I feel it's not fair to dismiss this problem as a waste of time.

**Reply by AmrSaber** [reply]
@JakubKoralewski That is exactly what I mean, For such a big framework, this is not just "some extra data that gets sent" and we need to optimize somewhere else, it's more like sending double the content of the page each time the page is loaded and that's for no apparent reason.
 
I -too- have no idea why doesn't that content gets removed once it's used to populate the page on the server, I don't know why the client needs to know about that data 🤷‍♂️ 

Hope someone fixes that at some point of time.

**Reply by timneutkens** [reply]
I assume you haven't looked into how React hydration / SSR with hydration works.

Here's some articles you'll want to read:
https://reactjs.org/docs/react-dom.html#hydrate
https://reactjs.org/docs/react-dom.html#render
https://joshwcomeau.com/react/the-perils-of-rehydration/
https://www.aboutmonica.com/blog/server-side-rendering-react-hydration-best-practices

In order to correctly hydrate and subsequently render the React tree (virtual DOM) the same props that were available server-side have to be available client-side as well. Meaning that the data that was used to render has to be available in the browser as well. Hence why the `__NEXT_DATA__` JSON script tag exists, it allows for passing the properties that were used during rendering to the browser so that React can be correctly hydrated. If this was not the case the React components would not have `props` client-side causing React to throw away the DOM that was rendered during pre-rendering. This throwing away of the DOM also happens when there is a mismatch between what was pre-rendered and what is rendered in the browser.

If you look at some of the largest websites / webapps in the world (by Alexa rank) you'll see that they use this technique even when not using Next.js.

Alexa #2 youtube.com - `window["ytInitialData"]`
Alexa #4 baidu.com - `hotsearch_data`
Alexa #5 facebook.com - `HasteSupportData`, `Bootloader` and much more
Alexa #6 qq.com - uses Next.js
Alexa #8 taobao.com - `J_ContextData`
Alexa #11 yahoo.com - `id="fc"`
Alexa #12 amazon.com - inlines snippets of data for each JS component
Alexa #18 reddit.com - `window.___r`
Alexa #20 netflix.com - `netflix.reactContext`
Alexa #25 instagram.com - `window._sharedData`
Alexa #27 csdn.net - `window.__INITIAL_STATE__`
Alexa #33 twitch.tv - `delete window.webpackJsonp,function(d)`

Note that I haven't gone through every site in the list and this is just an excerpt.
